pkg/controller/networkpolicy/validate.go

```diff
@@ -0,0 +1,199 @@
+		klog.V(2).Info("Validating CREATE request for ClusterNetworkPolicy")
+		if curCNP.Spec.Tier == "" || staticTierSet.Has(curCNP.Spec.Tier) {
+			// Empty Tier name corresponds to default Tier
+			break
+		}
+		if ok := v.tierExists(curCNP.Spec.Tier); !ok {
+			allowed = false
+			reason = fmt.Sprintf("tier %s does not exist", curCNP.Spec.Tier)
+		}
+	case admv1.Update:
+		// CNP "tier" must exist before referencing
+		klog.V(2).Info("Validating UPDATE request for CNP")
+		if curCNP.Spec.Tier == "" || staticTierSet.Has(curCNP.Spec.Tier) {
+			// Empty Tier name corresponds to default Tier
+			break
+		}
+		if ok := v.tierExists(curCNP.Spec.Tier); !ok {
+			allowed = false
+			reason = fmt.Sprintf("tier %s does not exist", curCNP.Spec.Tier)
+		}
+	case admv1.Delete:
+		// Delete of CNP have no validation
+		allowed = true
+	}
+	return reason, allowed
+}
+
+// validateTier validates the admission of a Tier resource
+func (v *NetworkPolicyValidator) validateTier(curTier, oldTier *secv1alpha1.Tier, op admv1.Operation) (string, bool) {
+	allowed := true
+	reason := ""
+	switch op {
+	case admv1.Create:
+		klog.V(2).Info("Validating CREATE request for Tier")
+		if len(v.networkPolicyController.tierInformer.Informer().GetIndexer().ListIndexFuncValues(PriorityIndex)) >= maxSupportedTiers {
+			return fmt.Sprintf("maximum number of Tiers supported: %d", maxSupportedTiers), false
+		}
+		// Tier priority must not overlap reserved tier's priority
+		if reservedTierPriorities.Has(curTier.Spec.Priority) {
+			return fmt.Sprintf("tier %s priority %d is reserved", curTier.Name, curTier.Spec.Priority), false
+		}
+		// Tier priority must not overlap existing tier's priority
+		trs, err := v.networkPolicyController.tierInformer.Informer().GetIndexer().ByIndex(PriorityIndex, strconv.FormatInt(int64(curTier.Spec.Priority), 10))
+		if err != nil || len(trs) > 0 {
+			return fmt.Sprintf("tier %s priority %d overlaps with existing Tier", curTier.Name, curTier.Spec.Priority), false
+		}
+	case admv1.Update:
+		// Tier priority updates are not allowed
+		klog.V(2).Info("Validating UPDATE request for Tier")
+		if curTier.Spec.Priority != oldTier.Spec.Priority {
+			allowed = false
+			reason = "update to Tier priority is not allowed"
+		}
+	case admv1.Delete:
+		klog.V(2).Info("Validating DELETE request for Tier")
+		if reservedTierNames.Has(oldTier.Name) {
+			reason = fmt.Sprintf("cannot delete reserved tier %s", oldTier.Name)
+			return reason, false
+		}
+		// Tier with existing CNPs cannot be deleted
+		cnps, err := v.networkPolicyController.cnpInformer.Informer().GetIndexer().ByIndex(TierIndex, oldTier.Name)
+		if err != nil || len(cnps) > 0 {
+			allowed = false
+			reason = fmt.Sprintf("tier %s is referenced by %d NetworkPolic(y)ies", oldTier.Name, len(cnps))
+		}
+	}
+	return reason, allowed
+}
+
+func (v *NetworkPolicyValidator) tierExists(name string) bool {
+	_, err := v.networkPolicyController.tierLister.Get(name)
```

Comment:
Question - lock is not required here?

Reply:
listers and informers are thread safe .. so not needed here